FFI:   implement  return 'const char*'  value via  pointer

Implement return string value by pointer.

```php
<?php

#ifndef KPHP
define('kphp', 0);
if (false)
#endif
  define('kphp', 1);

#ifndef KPHP
$lib =  FFI::load(__DIR__.'/str_return.h' );
if ( false )
#endif
{
        FFI::load(__DIR__.'/str_return.h' );
        $lib = FFI::scope("STRING");
}

$in = "in-str";
$out = FFI::new("const char*");
$lib->str_return($in, FFI::addr($out));
```

str_return.h
```c++
#define FFI_SCOPE "STRING"
#define FFI_LIB         "./str_return.so"
#define FFI_LIB_STATIC  "str_return.a"

void str_return(const char* in, const char** out);
```

str_return.cpp

```c++
#include <stdio.h>

extern "C" void str_return(const char* in, const char** out)
{
        printf("%s", in );
        *out = in;
}
```

Now: crash compile time:
<pre>
  0% [total jobs 15] [left jobs 15] [running jobs 0] [waiting jobs 4]
kphp_out/kphp/o_15/src_str_returnc3d1a42c2b5aa43.cpp: In function ‘Optional<bool> f$src_str_returnc3d1a42c2b5aa43()’:
kphp_out/kphp/o_15/src_str_returnc3d1a42c2b5aa43.cpp:45:42: error: no match for ‘operator=’ (operand types are ‘CDataPtr<char>’ and ‘CDataPtr<const char>’)
   45 |   v$out = ffi_new_cdata_ptr<const char*>();
      |                                          ^
In file included from /usr/share/vkontakte/kphp_source/objs/generated/auto/runtime/runtime-headers.h:16:
/usr/share/vkontakte/kphp_source/runtime/ffi.h:37:8: note: candidate: ‘constexpr CDataPtr<char>& CDataPtr<char>::operator=(const CDataPtr<char>&)’
   37 | struct CDataPtr {
</pre>

--- a/compiler/code-gen/vertex-compiler.cpp
+++ b/compiler/code-gen/vertex-compiler.cpp
@@ -185,8 +185,10 @@ void compile_ffi_php2c_conv(VertexAdaptor<op_ffi_php2c_conv> root, CodeGenerator
   const FFIType *conv_type = FFIRoot::get_ffi_type(root->c_type);
   const FFIType *type = conv_type->kind == FFITypeKind::Var ? conv_type->members[0] : conv_type;
   std::string tag;
+  // some tags have unboxed specializations, other types are specialized with CData templates;
+  // see ffi_tag<> specializations in ffi.h for more info
   if (type->is_void_ptr()) {
-    tag = "void*";
+    tag = type->members[0]->is_const() ? "const void*" : "void*";
   } else if (type->is_cstring()) {
     tag = "const char*";
   } else if (vk::any_of_equal(type->kind, FFITypeKind::Pointer, FFITypeKind::Struct, FFITypeKind::StructDef, FFITypeKind::Union, FFITypeKind::UnionDef)) {

--- a/compiler/inferring/type-data.cpp
+++ b/compiler/inferring/type-data.cpp
@@ -156,6 +156,9 @@ std::string TypeData::as_human_readable(bool colored) const {
     }
   }
 
+  if (ffi_const_flag()) {
+    res = "const " + res;
+  }
   if (indirection_ != 0) {
     res += std::string(indirection_, '*');
   }
@@ -216,6 +219,11 @@ void TypeData::set_ffi_pointer_type(const TypeData *new_ptr_type, int new_indire
   }
 
   auto *ptr_class = class_type()->ffi_class_mixin;
+
+  if (!ptr_class) {
+    set_ptype(tp_Error);
+    return;
+  }
 
   if (ptr_class->ffi_type->kind == FFITypeKind::Void && indirection_ == 1) {
     // any pointer is compatible with `void*`,
@@ -553,10 +561,14 @@ inline void get_cpp_style_type(const TypeData *type, std::string &res) {
         if (as_ffi->is_ref()) {
           res += klass->src_name;
         } else if (type->get_indirection() != 0) {
+          // indirection level (num of '*') and constness is qualifier-like property
+          // of a type, so FFI Type associated with a given class doesn't contain that info;
+          // therefore, it's not enough to just use ffi_mangled_decltype_string here -
+          // we need to add extra information using the TypeData state
           auto ptr = std::string(type->get_indirection() - 1, '*');
-          res += "CDataPtr<" + ffi_mangled_decltype_string(as_ffi->scope_name, FFIRoot::get_ffi_type(klass)) + ptr + ">";
+          std::string maybe_const = type->ffi_const_flag() ? "const " : "";
+          res += "CDataPtr<" + maybe_const + ffi_mangled_decltype_string(as_ffi->scope_name, FFIRoot::get_ffi_type(klass)) + ptr + ">";
         } else {
-          // TODO: can we avoid manual ptr addition here?
           res += "class_instance<C$FFI$CData<" + ffi_mangled_decltype_string(as_ffi->scope_name, FFIRoot::get_ffi_type(klass)) + ">>";
         }
         break;
@@ -865,6 +877,9 @@ bool is_less_or_equal_type(const TypeData *given, const TypeData *expected, cons
         break;
       case tp_Class:
         if (given->class_types() == expected->class_types()) {
+          if (given->ffi_const_flag() && !expected->ffi_const_flag()) {
+            return false;
+          }
           if (given->get_indirection() == expected->get_indirection()) {
             return true;
           }

--- a/compiler/inferring/type-data.h
+++ b/compiler/inferring/type-data.h
@@ -27,6 +27,7 @@ private:
     or_null_flag_e        = 0b00000010,
     or_false_flag_e       = 0b00000100,
     shape_has_varg_flag_e = 0b00001000,
+    ffi_const_flag_e      = 0b00010000,
   };
 
 public:
@@ -99,6 +100,9 @@ public:
   void set_flags(uint8_t new_flags);
 
   bool is_ffi_ref() const;
+
+  bool ffi_const_flag() const { return get_flag<ffi_const_flag_e>(); }
+  void set_ffi_const_flag() { set_flag<ffi_const_flag_e>(); }
 
   bool or_false_flag() const { return get_flag<or_false_flag_e>(); }
   void set_or_false_flag() { set_flag<or_false_flag_e>(); }

--- a/compiler/inferring/type-hint-recalc.cpp
+++ b/compiler/inferring/type-hint-recalc.cpp
@@ -111,8 +111,13 @@ void TypeHintCallable::recalc_type_data_in_context_of_call(TypeData *dst, Vertex
 
 static void recalc_ffi_type(TypeData *dst, const std::string &scope_name, const FFIType *type) {
   if (type->kind == FFITypeKind::Pointer) {
-    recalc_ffi_type(dst, scope_name, type->members[0]);
-    dst->set_indirection(type->num);
+    TypeData nested{*TypeData::get_type(tp_any)};
+    recalc_ffi_type(&nested, scope_name, type->members[0]);
+    nested.set_indirection(type->num); // turn it into a pointer
+    if (type->members[0]->is_const()) {
+      nested.set_ffi_const_flag();
+    }
+    dst->set_lca(&nested);
     return;
   }
 

--- a/compiler/phpdoc.cpp
+++ b/compiler/phpdoc.cpp
@@ -257,6 +257,8 @@ const TypeHint *PhpDocTypeHintParser::parse_ffi_cdata() {
       cdef.append("int");
     } else if (tok == tok_times) {
       cdef.push_back('*');
+    } else if (tok == tok_const) {
+      cdef.append("const");
     } else {
       throw std::runtime_error("unexpected token, expected C type, '>' or ')'");
     }

--- a/compiler/pipes/final-check.cpp
+++ b/compiler/pipes/final-check.cpp
@@ -309,6 +309,9 @@ bool is_php2c_valid(VertexAdaptor<op_ffi_php2c_conv> conv, const FFIType *ffi_ty
       if (php_expr->type() == op_null) {
         return true;
       }
+      if (!ffi_type->members[0]->is_const() && php_type->ffi_const_flag()) {
+        return false;
+      }
       if (ffi_type->members[0]->kind == FFITypeKind::Void && ffi_type->num == 1) {
         return true;
       }

--- a/docs/kphp-language/php-extensions/ffi.md
+++ b/docs/kphp-language/php-extensions/ffi.md
@@ -53,7 +53,7 @@ All FFI-related types should be annotated with special annotations:
 * `ffi_cdata<scope_name, type_expr>` for [\FFI\CData](https://www.php.net/manual/ru/class.ffi-cdata.php)
 * `ffi_scope<scope_name>` for [\FFI](https://www.php.net/manual/ru/class.ffi.php) class instances
 
-For `ffi_scope<>`, you can use a pointer-type expression, builtin type name, struct name and
+For `ffi_cdata<>`, you can use a pointer-type expression, builtin type name, struct name and
 any other supported C type expression. For instance, `ffi_cdata<foo, union Event**>` is a valid
 `ffi_cdata` declaration.
 
@@ -82,6 +82,21 @@ $cdef = FFI::cdef('
 ');
 
 f($cdef);
+```
+
+If you need to describe a builtin C type that is not bound to any particular FFI scope,
+use default FFI scope name "C":
+
+```php
+/**
+ * Since void* is a global type, we can use "C" scope name to avoid any dependency
+ * on a custom FFI scope.
+ *
+ * @param ffi_cdata<C, void*> $ptr
+ */
+function as_uint8_ptr($ptr) {
+  FFI::cast('uint8_t*', $ptr);
+}
 ```
 
 KPHP will give compile-time error if types are not compatible

--- a/runtime/ffi.h
+++ b/runtime/ffi.h
@@ -51,6 +51,12 @@ struct CDataPtr {
     return c_value == reinterpret_cast<T*>(1);
   }
 
+  // allow `T*` ptr to `void*` ptr conversion
+  operator CDataPtr<void> () { return CDataPtr<void>(c_value); }
+
+  // allow `T*` ptr to `const T*` ptr conversion
+  operator CDataPtr<const T> () { return CDataPtr<const T>(c_value); }
+
   // construction FFI pointer from KPHP null value
   CDataPtr(Optional<bool> v) {
     if (!v.is_null()) {
@@ -316,6 +322,7 @@ template<> struct ffi_tag<float>{};
 template<> struct ffi_tag<double>{};
 template<> struct ffi_tag<char>{};
 template<> struct ffi_tag<const char*>{};
+template<> struct ffi_tag<const void*>{};
 template<> struct ffi_tag<void*>{};
 template<class T> struct ffi_tag<C$FFI$CData<T>>{};
 
@@ -331,12 +338,16 @@ inline int64_t ffi_php2c(int64_t v, ffi_tag<int64_t>) { return v; }
 inline float ffi_php2c(double v, ffi_tag<float>) { return static_cast<float>(v); }
 inline double ffi_php2c(double v, ffi_tag<double>) { return v; }
 inline const char* ffi_php2c(const string &v, ffi_tag<const char*>) { return v.c_str(); }
+inline const void* ffi_php2c(const string &v, ffi_tag<const void*>) { return v.c_str(); }
 
 template<class T>
 auto ffi_php2c(class_instance<C$FFI$CData<T>> v, ffi_tag<C$FFI$CData<T>>) { return v->c_value; }
 
 template<class T>
 void *ffi_php2c(CDataPtr<T> v, ffi_tag<void*>) { return v.c_value; }
+
+template<class T>
+const void *ffi_php2c(CDataPtr<const T> v, ffi_tag<const void*>) { return v.c_value; }
 
 template<class T>
 T* ffi_php2c(CDataPtr<T> v, ffi_tag<C$FFI$CData<T*>>) { return v.c_value; }

--- a/tests/phpt/ffi/typing/010_cstr_field_read_error.php
+++ b/tests/phpt/ffi/typing/010_cstr_field_read_error.php
@@ -1,6 +1,6 @@
 @kphp_should_fail
 KPHP_ENABLE_FFI=1
-/pass FFI\\CData_char\* to argument/
+/pass const FFI\\CData_char\* to argument/
 /declared as @param string/
 <?php
 

--- a/tests/phpt/ffi/typing/054_const_ptr_error.php
+++ b/tests/phpt/ffi/typing/054_const_ptr_error.php
@@ -1,0 +1,11 @@
+@kphp_should_fail
+KPHP_ENABLE_FFI=1
+/Invalid php2c conversion: const FFI\\CData_int32\* -> int32_t\*/
+<?php
+
+$cdef = FFI::cdef('
+  void f_intptr(int*);
+');
+
+$const_intptr = FFI::new('const int*');
+$cdef->f_intptr($const_intptr);

--- a/tests/phpt/ffi/typing/055_const_ptr_error.php
+++ b/tests/phpt/ffi/typing/055_const_ptr_error.php
@@ -1,0 +1,16 @@
+@kphp_should_fail
+KPHP_ENABLE_FFI=1
+/pass const FFI\\CData_int32\* to argument \$x of f/
+/but it's declared as @param FFI\\CData_int32\*/
+<?php
+
+// Passing `const T*` as `T*` is a type error.
+
+/** @param ffi_cdata<C, int32_t*> $x */
+function f($x) {}
+
+$intptr = FFI::new('int32_t*');
+$const_intptr = FFI::new('const int32_t*');
+
+f($const_intptr);
+f($intptr);

--- a/tests/phpt/ffi/typing/056_const_ptr_error.php
+++ b/tests/phpt/ffi/typing/056_const_ptr_error.php
@@ -1,0 +1,14 @@
+@kphp_should_fail
+KPHP_ENABLE_FFI=1
+/pass const FFI\\CData_int32\* to argument \$x of f/
+/but it's declared as @param Foo/
+<?php
+
+class Foo {}
+
+/** @param Foo $x */
+function f($x) {}
+
+$const_intptr = FFI::new('const int32_t*');
+
+f($const_intptr);

--- a/tests/python/tests/ffi/c/pointers.c
+++ b/tests/python/tests/ffi/c/pointers.c
@@ -42,3 +42,15 @@ void bytes_array_set(uint8_t *arr, int offset, uint8_t val) {
 void* ptr_to_void(void *ptr) {
   return ptr;
 }
+
+const void* ptr_to_const_void(const void *ptr) {
+  return ptr;
+}
+
+int void_strlen(const void *s) {
+  return strlen(s);
+}
+
+void cstr_out_param(const char **out) {
+  *out = "example result";
+}

--- a/tests/python/tests/ffi/c/pointers.h
+++ b/tests/python/tests/ffi/c/pointers.h
@@ -21,3 +21,7 @@ uint8_t bytes_array_get(uint8_t *arr, int offset);
 void bytes_array_set(uint8_t *arr, int offset, uint8_t val);
 
 void* ptr_to_void(void *ptr);
+const void* ptr_to_const_void(const void *ptr);
+int void_strlen(const void *s);
+
+void cstr_out_param(const char **out);

--- a/tests/python/tests/ffi/php/pointer/cast.php
+++ b/tests/python/tests/ffi/php/pointer/cast.php
@@ -185,9 +185,86 @@ function test_cast_struct_non_pointer() {
   var_dump([$as_foo2->value, $foo->value, $single_int_wrapper->single_int->x]);
 }
 
+/**
+  * @param ffi_cdata<C, void*> $ptr
+  */
+function as_byte_ptr($ptr) {
+  return FFI::cast('uint8_t*', $ptr);
+}
+
+function test_cast_via_func() {
+  $lib = FFI::scope('pointers');
+
+  $int64 = FFI::new('int64_t');
+  $as_bytes = as_byte_ptr(FFI::addr($int64));
+  for ($i = 1; $i < 8; $i++) {
+    $lib->bytes_array_set($as_bytes, $i, $i * 4);
+    var_dump([
+      $lib->bytes_array_get($as_bytes, $i-1),
+      $lib->bytes_array_get($as_bytes, $i),
+    ]);
+  }
+
+  $int32 = FFI::new('int32_t');
+  $as_bytes = as_byte_ptr(FFI::addr($int32));
+  for ($i = 1; $i < 4; $i++) {
+    $lib->bytes_array_set($as_bytes, $i, $i * 4);
+    var_dump([
+      $lib->bytes_array_get($as_bytes, $i-1),
+      $lib->bytes_array_get($as_bytes, $i),
+    ]);
+  }
+}
+
+/**
+  * @param ffi_cdata<C, const void*> $ptr
+  */
+function as_const_byte_ptr($ptr) {
+  return FFI::cast('const uint8_t*', $ptr);
+}
+
+function test_string_to_void_pointer() {
+  $lib = FFI::scope('pointers');
+
+  $str = "hello";
+  $void_ptr = $lib->ptr_to_const_void($str);
+  var_dump($lib->void_strlen($void_ptr));
+  var_dump($lib->void_strlen($str));
+  var_dump($lib->void_strlen(as_const_byte_ptr($void_ptr)));
+}
+
+// This function is called with `int64_t*` and `const int64_t*`,
+// it infers `const int64_t*` as lca type and accepts them both
+function auto_infer_const_ptr($x) {}
+
+/** @param ffi_cdata<C, const int64_t*> $x */
+function accept_const_int64_ptr($x) {}
+
+function test_const_cast() {
+  $lib = FFI::scope('pointers');
+
+  $i64 = FFI::new('int64_t');
+  $intptr = FFI::addr($i64);
+  $const_intptr = FFI::cast('const int64_t*', $intptr);
+
+  accept_const_int64_ptr($const_intptr);
+  accept_const_int64_ptr($intptr);
+  var_dump($lib->read_int64($const_intptr));
+  var_dump($lib->read_int64($intptr));
+  $i64->cdata = 13;
+  var_dump($lib->read_int64($const_intptr));
+  var_dump($lib->read_int64($intptr));
+
+  auto_infer_const_ptr($const_intptr);
+  auto_infer_const_ptr($intptr);
+}
+
 test_cast_to_self();
 test_cast_void_pointer();
 test_cast_scalar_pointer();
 test_cast_scalar_non_pointer();
 test_cast_struct_pointer();
 test_cast_struct_non_pointer();
+test_cast_via_func();
+test_string_to_void_pointer();
+test_const_cast();

--- a/tests/python/tests/ffi/php/pointer/ptr2.php
+++ b/tests/python/tests/ffi/php/pointer/ptr2.php
@@ -37,6 +37,14 @@ function test() {
     var_dump($lib->voidptr_addr_value($ptr1) === $lib->voidptr_addr_value($ptr2));
 }
 
+function test_ptr2_out_param() {
+  $lib = FFI::scope('pointers');
+
+  $out_ptr = FFI::new('const char*');
+  $lib->cstr_out_param(FFI::addr($out_ptr));
+  var_dump(FFI::string($out_ptr));
+}
+
 function test2() {
     $lib = FFI::scope('pointers');
     $u64 = $lib->new('uint64_t');
@@ -50,3 +58,4 @@ function test2() {
 
 test();
 test2();
+test_ptr2_out_param();

